src/librustc/traits/select.rs

```diff
@@ -617,6 +625,17 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         })
     }
 
+    /// Evaluates whether the obligation `obligation` can be satisfied and returns
+    /// an `EvaluationResult`.
+    pub fn evaluate_obligation_recursively(&mut self,
```

Comment:
Not sure if exposing this new API is a good thing to have done, but it was the cleanest way without also exposing `TraitObligationStackList` (which is even more of an implementation detail).

Reply:
Seems OK -- I might prefer to remove `evaluate_obligation_conservatively` and `evaluation_obligation` in favor of just offering this, though. Or at least have them invoke it.

Reply:
Good point. The only caller of the originals now is in coherence, so I'll just switch that to use this and remove both the originals. And rename this to just `evaluate_obligation` to boot.